### ISSUE TITLE
Add Purged K-Fold Cross Validation Class

### ISSUE THREAD

**Issue by Patschkowski** [mlpack/mlpack#3830]
### What is the desired addition or change?

It is envisioned to add a third Cross Validation (CV) class that purges and embargos data from the training sets after the k-fold splits to avoid information leakage from the training sets into the test set for time series. Algorithm will remove data from the training set as per the following picture ("overlap" and "embargo" will be removed).

![image](https://github.com/user-attachments/assets/14280d83-a8e2-4af7-beff-2acc533010a4)

### What is the motivation for this feature?

Time series are not IID. New samples depend on previous samples and thus information is leaked from the training sets into the test set if not prevented.

### If applicable, describe how this feature would be implemented.

mlpack has two implementations of CV: KFoldCV and SimpleCV. Both use functionality from CVBase. Proposed is to add a third class PurgedKFoldCV that implements the above-mentioned functionality.
 
### Additional information?

Literature reference: M. López de Prado (2018): "Advances in Financial Machine Learning", pp. 105-109

**Comment by dikshant182004** [comment]
hii , @Patschkowski i have read your issue and wanna contribute in this issue . i have read the reference thet u mentioned .so  think we can implement this by creating a different class where we will pass embargo and purge as an argument which will be used to avoid data leakage from the training set  into test set .
we can create a purge window which will remove data from the end of the train set and a embargo window which will block some data points just before the test data starts.(we can define them in percentage so it gets adjusted based on the sample size).so in this way we can creata a significant time difference between train and test set preventing data leakage

**Comment by dikshant182004** [comment]
however i thought there is problem with this issue since the difference we are creating between train test and test set can lead to decrease in data which may result in high bias

**Comment by Patschkowski** [comment]
@dikshant182004 , great to hear that you also want to contribute. Why do you think it will result in a higher bias?

**Comment by dikshant182004** [comment]
no it will not i forget we are just evaluating the model

**Comment by dikshant182004** [comment]
since there is decrease in training data .so may be it will not be able to find the underling patterns in training data well that may result in high bias

**Comment by dikshant182004** [comment]
since we are just evaluating its accuracy so may be we should not concrened more about this

**Comment by dikshant182004** [comment]
hi @Patschkowski any suggestion on my appoach so that i can proceed furthur

**Comment by Patschkowski** [comment]
Hi @dikshant182004 my understanding of the reference is, that indeed for the embargo it works on percentages (p. 108) but for the purging it is based on the beginning and ending of the observations (p. 106). I have not studied the details yet, but I think it is very similar to the indicator matrix concept introduced in a previous chapter (p. 64). How do you look at it?

I was also thinking about the design. Either we create a new class and use CVBase as the SimpleCV and KFoldCV classes are doing. Or we make KFoldCV configurable to perform purging/embargoing as well, or we make KFoldCV a base class and derive PurgedKFoldCV from it. We could use static polymorphism to overwrite the functionality in KFoldCV from PurgedKFoldCV. I think that will be the most elegant approach and we do not need to replicate so much code and only slightly modify KFoldCV but need to introduce CRTP.

**Comment by Sonu0305** [comment]
Hi @rcurtin & @Patschkowski ,
I came across this issue regarding the addition of a third Cross Validation (CV) class for time series data, specifically the `PurgedKFoldCV` to prevent information leakage by purging and embargoing data after the k-fold splits.

I think this is an important feature, and I’m interested in contributing by implementing the new class. Before I begin, I wanted to clarify a few things:
1. We could introduce the new class `PurgedKFoldCV`, as proposed, which would extend `CVBase` and implement the purging and embargoing behavior separately.
2. Alternatively, we could modify the existing `KFoldCV` to make it configurable for purging/embargoing, or derive `PurgedKFoldCV` directly from `KFoldCV` to avoid code duplication.

Which approach would you prefer, or do you have any specific design suggestions?
If currently no one is working on this, I’d be happy to start working on this and submit a PR once the implementation is ready. Looking forward to your feedback and guidance, Thank You.

**Comment by Patschkowski** [comment]
@Sonu0305 , let us wait until #3829 is merged. That also deals with the indicator matrix needed for the purging. Once that is accepted we can move ahead with this one.

**Comment by Sonu0305** [comment]
Thanks for the update, @Patschkowski. I understand, I’ll wait for #3829 to be merged before proceeding with the implementation of `PurgedKFoldCV`. Once it’s accepted, I’ll be ready to move forward with the work on this. Let me know if there’s anything I can help with in the meantime!

**Comment by Sonu0305** [comment]
Thanks for the suggestion, @Patschkowski! I’ll start working on the unit tests for the PurgedKFoldCV implementation in the meantime. I agree, it makes sense to get the tests in place first so we can easily integrate them once the final implementation is settled. Let me know if you have any specific requirements or guidance for the tests.

**Comment by Darshit324** [comment]
hi guys i too want to contribute in this issue, though im just a beginner ill try to help in any way i can i looked into issue and i think i got the basic understanding of it, ill try changing a few lines

**Comment by Darshit42** [comment]
hi guys i too want to contribute in this issue, though im just a beginner ill try to help in any way i can i looked into issue and i think i got the basic understanding of it, ill try changing a few lines

**Comment by Patschkowski** [comment]
@Darshit42 , @Sonu0305 , @dikshant182004 ,

let us move ahead and implement this one. I'll share my thoughts later today on how to set it up and distribute the work.

@rcurtin ,

hope you can also give us some guidance on the proposal once available and guide us through the process as you did previously with the other PR!!

**Comment by Darshit42** [comment]
guys is anyone of you on slack or discord? @Patschkowski , @dikshant182004 , @Sonu0305 im having some issue while i was testing the code so i completely removed my entire local setup and reinstalled it , now im facing issue in local setup :(

**Comment by dikshant182004** [comment]
@Darshit42 tell me about the issue u are facing or we can talk on slack regarding that

**Comment by dev484p** [comment]
I'm interested in contributing and wanted to check in on the current status of things. Are there any open tasks or areas where I could jump in and help?

I'm also a beginner-intermediate in this domain.

**Comment by Shreyanotdeadyet** [comment]
I’d like to work on this issue regarding implementing PurgedKFoldCV for time series cross-validation. I’ve gone through the reference and the discussions in this thread. Could you let me know the current status of this issue? Additionally, is there a Discord or other communication channel where I can stay updated and collaborate more effectively?

Proposed Approach:

Introducing a new PurgedKFoldCV class inheriting from CVBase, similar to KFoldCV and SimpleCV, to implement purging (removing overlaps) and embargoing (blocking buffer zones) as configurable parameters. Starting with unit tests using synthetic time series data to validate the behavior for varying purgeWindow and embargoWindow values, integrate the indicator matrix from issue #3829, and use CRTP to minimize code duplication by reusing functionality from KFoldCV.

**Comment by Patschkowski** [comment]
To be handled in #3909
